content/blog/2022/2022-08-30-handling-timezones-in-spring.md

```diff
@@ -0,0 +1,516 @@
+
+### Mutable Classes
+
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/mutableDate.JPG" %}}
+
+- Immutability is a key concept that ensures that java objects are thread-safe and concurrent access does not lead to an inconsistent state.
+- The Date API **allows mutable methods** such as `setHours`, `setMinutes`, `setDate`. Therefore, it becomes the **responsibility
+of the developer to clone the object before use**.
+- Similarly, the Calendar object also has setter methods `setTimeZone`, `add` which allows an object to be modified.
+
+### Format Dates
+
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/dateFormat.JPG" %}}
+
+With the Date API, formatting can be quite tedious and the process involves numerous steps. 
+As seen in the example above, there are various flaws in this process:
+- The Date API itself does not store any formatting information. Therefore, we need to use it in combination with the `SimpleDateFormat`.
+- The `SimpleDateFormat` class is not thread-safe so it **cannot be used in multithreaded applications without proper synchronization**.
+- As the Date API **does not have timezone information**, we have to use the `Calendar` class. However, it cannot be formatted, 
+so we extract date from `Calendar` for formatting.
+
+#### SQL Dates
+
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/sqlDates.JPG" %}}
+
+- The API demonstrates poor design choice as `java.sql.Date`, `java.sql.Time` and `java.sql.Timestamp` all extend `java.util.Date` class.
+Due to differences between the subclasses and java.util.Date, the [documentation](https://docs.oracle.com/javase/7/docs/api/java/sql/Timestamp.html) itself suggests to not use 
+the Date class generically thus violating the **Liskov Substitution Principle.**
+
+{{% info title="Deprecated methods" %}}
+- Most of the methods in the `java.util.Date` class are [deprecated](https://docs.oracle.com/javase/7/docs/api/deprecated-list.html#method). However, they are not officially removed from the JDK library to support legacy applications.
+- To overcome the shortcomings of java.util classes, Java8 introduced the new **DateTime API** in the `java.time` package. 
+{{% /info %}}
+
+## Java8 DateTime API
+
+The [DateTime API](https://docs.oracle.com/javase/8/docs/api/java/time/package-summary.html) is heavily influenced by the [Jodatime](https://www.joda.org/joda-time/) library which was the defacto standard prior to Java8.
+In this section, we will look at some commonly used date-time classes and its corresponding operations in this section.
+
+### LocalDate
+
+**[java.time.LocalDate](https://docs.oracle.com/javase/8/docs/api/java/time/LocalDate.html)** is an immutable date object that does not store time or timezone information. However, we can pass the `java.time.ZoneId` object to get the local date in a particular timezone.
+
+Sample conversion examples: 
+
+````java
+
+        LocalDate today = LocalDate.now();
```

Comment:
This and the following code snippets are worth gold for people who are searching for conversions between date objects!

However, it's hard to follow all the `System.out.println()`s. Can you convert the code examples to unit tests with assertions? This way, the reader sees an assertions like `assertThat(today).hasDayOfMonth(12)`. If you use assertJ as the assertion library (like in the example) the tests become very readable and the reader gets the most value (and learns that assertJ is a cool library to test dates).

Also, you wouldn't need to show any console output at all, because the tests are self-explanatory.